overlord/servicestate/quota_handlers.go

```diff
@@ -0,0 +1,262 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package servicestate
+
+import (
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/progress"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/snap/quota"
+	"github.com/snapcore/snapd/snapdenv"
+	"github.com/snapcore/snapd/strutil"
+	"github.com/snapcore/snapd/systemd"
+	"github.com/snapcore/snapd/timings"
+	"github.com/snapcore/snapd/wrappers"
+)
+
+type ensureSnapServicesForGroupOptions struct {
+	// extraSnaps is the set of extra snaps to consider when ensuring services,
+	// mainly only used when snaps are removed from quota groups
+	extraSnaps []string
+
+	// meter is the progress meter used by the systemd package for reporting
+	// changes and progress
+	meter progress.Meter
+
+	// perfTimings is the timings for this task
+	perfTimings *timings.Timings
```

Comment:
I support the idea of introducing options struct, in particular if we expect the number of options to grow, but I find passing meter and perfTimings via it very unusual, we haven't done this before. I'd leave them out and keep ensureSnapServicesForGroupOptions limited to quota-related options.

Reply:
yes, this is unusual

Reply:
I'm trying to avoid this sort of signature:
```go 
func ensureSnapServicesForGroup(st *state.State, grp *quota.Group, allGrps map[string]*quota.Group, extraSnaps []string, meter progress.Meter perfTimings *timings.Timings) error {
```

what would you suggest as a better signature? I could put `grp`, `allGrps` and `extraSnaps` into an options struct I suppose but that also feels weird and in some cases this function becomes called as

```go
type ensureOptions struct {
    grp *quota.Group
    allGrps map[string]*quota.Group
    extraSnaps []string
} 

func ensureSnapServicesForGroup(st *state.State, meter progress.Meter perfTimings *timings.Timings, opts *ensureOptionsStruct) error {
   ...
}
func foo() {
    ensureSnapServicesForGroup(st, progress.Null, nil, opts)
}
```

which does not read well to me

Reply:
FWTW that's not too far from what we have in other places, and it is ok from consistency pov.

I'd however make progress and timings last as they are the least important arguments. And maybe pass group explicitly as it's the subject of the function. So either:

`func ensureSnapServicesForGroup(st *state.State, opts *ensureOptionsStruct, meter progress.Meter perfTimings *timings.Timings) error`

or (that would be my personal preference):

`
func ensureSnapServicesForGroup(st *state.State, grp *quota.Group, opts *ensureOptionsStruct, meter progress.Meter perfTimings *timings.Timings) error`

(with the latter only allGrps and extraSnaps go into ensureOptionsStruct)

Reply:
I agree with @stolowski preference here if that suggestion makes sense

Reply:
I went with the first suggestion, it's a bit unfortunate since the opts can't be nil now but oh well

Reply:
I'm nor sure why you didn't pick the 2nd though, the one we both said was the actual preference

Reply:
I guess I misread the sentence to mean that the first option was preferred: `(that would be my personal preference)` considering `that` is past tense I assumed y'all were referring to the first proposal that Pawel had...

in any case I don't really see how the second one is better than just having all the arguments direct without any options - considering the caller still always has to pass in opts since we always need `allGrps`, so essentially opts is just being used for `extraSnaps` so what's the point ... at the end of the day though this function is not exported so the signature is pretty pointless I think so I will just go with what you wanted anyways to try and land this

Reply:
@anonymouse64 sorry if my comment was confusing - yes, my preference was the 2nd one. As I said in my first comment  opts make sense if we expect it to grow (I don't know if this is the case in the upcoming PRs), so maybe it is borderline right now; in any case I'm fine with it as is, thanks for the changes!